### PR TITLE
Removed support for writing denormalised tokens in the ring.

### LINKED ISSUE #1877: Inconsistent behaviour of `Desc.AddIngester`
Before the introduction of normalized tokens, `Desc.AddIngester` used to append tokens to the existing ones. But currently, it works as `append` when it's not normalized tokens and `replace` when it's normalized tokens.

https://github.com/cortexproject/cortex/blob/08ddf88e082352a89a6daa08fad6c7064d7a731a/pkg/ring/model.go#L38-L63

Looking at places where `Desc.AddIngester` is used, `append` seems to be the desired behaviour. Discovered in #1750 and @pstibrany plans to fix it in #1809

### PR DESCRIPTION
**Pull request by pstibrany**
Removed support for writing denormalised tokens in the ring.

Still supported:
- claiming tokens from another ingester, that is using denormalised tokens
- finding denormalised tokens in the ring, when joining

Other ingesters can still use denormalised tokens, migrateRing takes care of that.

Targeted for Cortex 0.5.0. ~No changelog written yet.~

Issue #1787
Fixes #1877

### CODE REVIEW

File: docs/arguments.md [hunk 1]
```diff
@@ -234,9 +234,9 @@ It also talks to a KVStore and has it's own copies of the same flags used by the
 
 - `-ingester.normalise-tokens`
 
-   Write out "normalised" tokens to the ring.  Normalised tokens consume less memory to encode and decode; as the ring is unmarshalled regularly, this significantly reduces memory usage of anything that watches the ring.
+   Deprecated. New ingesters always write "normalised" tokens to the ring. Normalised tokens consume less memory to encode and decode; as the ring is unmarshalled regularly, this significantly reduces memory usage of anything that watches the ring.
```
Comment (pracucci) [inline docs/arguments.md]: This change should be also clearly stated out in the changelog.

File: pkg/ring/lifecycler.go [hunk 2]
```diff
@@ -82,7 +82,7 @@ func (cfg *LifecyclerConfig) RegisterFlagsWithPrefix(prefix string, f *flag.Flag
 	f.DurationVar(&cfg.ObservePeriod, prefix+"observe-period", 0*time.Second, "Observe tokens after generating to resolve collisions. Useful when using gossiping ring.")
 	f.DurationVar(&cfg.MinReadyDuration, prefix+"min-ready-duration", 1*time.Minute, "Minimum duration to wait before becoming ready. This is to work around race conditions with ingesters exiting and updating the ring.")
 	flagext.DeprecatedFlag(f, prefix+"claim-on-rollout", "DEPRECATED. This feature is no longer optional.")
-	f.BoolVar(&cfg.NormaliseTokens, prefix+"normalise-tokens", false, "Store tokens in a normalised fashion to reduce allocations.")
+	flagext.DeprecatedFlag(f, prefix+"normalise-tokens", "DEPRECATED. This feature is no longer optional")
```
Comment (pracucci) [inline pkg/ring/lifecycler.go]: Add a `.` at the end of the description sentence, please.

File: pkg/ring/model.go [hunk 3]
```diff
@@ -45,18 +45,17 @@ func (d *Desc) AddIngester(id, addr string, tokens []uint32, state IngesterState
...
 
-	if normaliseTokens {
-		ingester.Tokens = tokens
-	} else {
-		for _, token := range tokens {
-			d.Tokens = append(d.Tokens, TokenDesc{
-				Token:    token,
-				Ingester: id,
-			})
+	// delete any reference from d.Tokens (can happen if denormalized tokens for this ingester
```
Comment (pracucci) [inline pkg/ring/model.go]: I'm missing some context. Can you help me understand this change, so that I can make a more conscious review, please?

File: docs/arguments.md (hunk 1 above)
Comment (pstibrany) [reply docs/arguments.md]: I don't want to update changelog yet, as 0.4.0 is going to be released soon (13th nov), and then I'll need to change it anyway. But you're right, this needs to go to changelog.

File: pkg/ring/model.go (hunk 3 above)
Comment (pstibrany) [reply pkg/ring/model.go]: See pkg/ring/lifecycler.go:549, autoJoin method.

When ingester finds that it already owns some tokens during join, it will use those tokens, instead of generating new ones. Those tokens could have been normalized or not. If they were not normalized (i.e. stored in d.Tokens), we need to remove them, otherwise they would be stored at two places (denormalized in `d.Tokens`, but also normalized tokens in `Ingesters[id].Tokens`).

Comment (pstibrany) [reply pkg/ring/model.go]: Also, this behavior is used by updated test.

Comment (pracucci) [inline pkg/ring/model.go]: I would rephrase the comment for clarity, explaining we do `delete any denormalized token for this ingester, given all ingester tokens are going to be stored normalized. It may happen to find denormalized tokens when joining the ring (ie. the previous instance of the same ingester was running with denormalized tokens).`

Comment (pracucci) [reply pkg/ring/model.go]: Right, got it, thanks!

Comment (pstibrany) [reply pkg/ring/model.go]: I've updated comment, and also added test for this. It turns out that my previous test was checking different scenario (takeover of denormalised tokens by different ingester), and commenting out this code didn't make it fail.

File: docs/arguments.md (hunk 1 above)
Comment (pstibrany) [reply docs/arguments.md]: Added changelog entry now.

File: pkg/ring/model.go [hunk 4]
```diff
@@ -45,18 +45,18 @@ func (d *Desc) AddIngester(id, addr string, tokens []uint32, state IngesterState
 		Addr:      addr,
 		Timestamp: time.Now().Unix(),
 		State:     state,
+		Tokens:    tokens,
```
Comment (codesome) [inline pkg/ring/model.go]: Created an issue for this just to not miss out: https://github.com/cortexproject/cortex/issues/1877

Comment (pstibrany) [reply pkg/ring/model.go]: I've added comment to AddIngester about tokens being replaced and test to make sure it will not break.

**Comment by pstibrany** [comment]
@gouthamve why 'type/breaking' label? We know that upgrade to normalised tokens works fine, and so making them mandatory won't break anything. (Perhaps if someone is upgrading from VERY old Cortex versions that didn't support normalisation at all?)

File: pkg/ring/lifecycler.go [hunk 5]
```diff
@@ -50,7 +50,7 @@ type LifecyclerConfig struct {
 	JoinAfter        time.Duration `yaml:"join_after,omitempty"`
 	MinReadyDuration time.Duration `yaml:"min_ready_duration,omitempty"`
 	UnusedFlag       bool          `yaml:"claim_on_rollout,omitempty"` // DEPRECATED - left for backwards-compatibility
```
Comment (gouthamve) [inline pkg/ring/lifecycler.go]: Can we move the unused flags to bottom of the struct definition?

File: CHANGELOG.md [hunk 6]
```diff
@@ -11,6 +11,7 @@
   * `ruler.poll-interval` has been added to specify the interval in which to poll new rule groups.
 * [FEATURE] The distributor can now drop labels from samples (similar to the removal of the replica label for HA ingestion) per user via the `distributor.drop-label` flag. #1726
 * [BUGFIX] Fixed unnecessary CAS operations done by the HA tracker when the jitter is enabled. #1861
+* [CHANGE] Ingesters now write only normalised tokens to the ring, although they can still read denormalised tokens used by other ingesters. `-ingester.normalise-tokens` is now deprecated, and ignored. If you want to switch back to using denormalised tokens, you need to downgrade to Cortex 0.4.0. Previous versions don't handle claiming tokens from normalised ingesters correctly. #1809
```
Comment (gouthamve) [inline CHANGELOG.md]: Change should be on top.

File: pkg/ring/lifecycler.go (hunk 5 above)
Comment (pstibrany) [reply pkg/ring/lifecycler.go]: Done.

File: CHANGELOG.md (hunk 6 above)
Comment (pstibrany) [reply CHANGELOG.md]: Moved.
